rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+
+
+
+*   P1 depends on P2
+*   P2 is expected to provide a public API
+*   All API symbols exposed by P2 version M.N is expected to work at version M.(N+K) for any non-negative integer K.
+*   P2 is allowed to make breaking changes to its API between major releases (M to M+1)
+*   If P1 version X.Y works with P2 version M.N, it should also work the same way with P2 version M.(N+K) However, there are no guarantees for it to work with P2 version (M+K).L
+*   When P1 is releasing a new version, it should check which API symbols it needs from P2, and fix the minimum version requirement in its pip package for P2 accordingly.
+
+
+### Optional TF packages
+
+Mostly expected to contain the C++ plugins defined in the previous section. These will be simple pip packages that will deploy the shared objects under "site-packages/tensorflow-plugins"
+
+These shared objects will be automatically loaded by TF core if:
+
+
+
+*   They correctly define the compatibility strings using `TF_PLATFORM_STRINGS`
+*   They are compatible with the system tf core is running on
+*   They have been properly built and signed (unless running in developer mode)
+
+
+## Alternatives / Potential Issues
+
+
+
+*   **Why do we not use C++ APIs instead of C**: Compilers have no guarantees for ABIs generated for C++ code. Any C++ API used will require each shared object to be compiled with the same compiler, using the same version of the compiler, with the same compiler flags ([See github issue 23561](https://github.com/tensorflow/tensorflow/issues/23561)).
+*   **Why do not we statically link everything**: Single shared object for everything: Anywhere except google does not have access to the massively parallel build system we use here at google. This causes prohibitive build times, causing major developer pain for open source developers. There are many more issues, but the summary is while this is a great solution for google, outside google this is simply infeasible.
+*   **TF will become a suite of multiple packages, built by multiple authorities. What if the bugs get blamed on TF team**: With the modular model, we expect testing of 3rd party code to become easier. This can also be mitigated if the error messages are better, and if they can clearly point out which module the issue stems from. Finally, we can create an apple-swift like testing model, where we run a Jenkins setup that people can donate their machines to, and we can run continuous integration tests on their plugins.
+*   **Why not have APIs but still have a monolithic repository: **When everything is in a single repository, this enables developers to bypass the APIs, and depend on internals. Moreover, we cannot grant full control over different folders on our repository to our partners in a single repository. As long as they are in a single repository, they are still constrained by our build system and license. Finally, in a single repository we do not provide the option of closed source plugins for contributors.
```

Comment:
What will be the constraint in regard of the licences? I guess if a module is outside the TF organization, license could be any that complies with the one with TensorFlow but what about modules that remains under github.com/tensorflow but outside the main repo?

Reply:
https://github.com/tensorflow/community/pull/18#issuecomment-424389704